### PR TITLE
allow metaclass keywords in Model subclasses

### PR DESCRIPTION
**Pull request by nstarman**
Signed-off-by: Nathaniel Starkman (@nstarman) <nstarkman@protonmail.com>

### Description

In the Model class construction the metaclass ``_ModelMeta`` does not permit keyword arguments. Adding this option does not change anything in the class creation process, merely allows subclasses of ``_ModelMeta`` to inherit methods without having to wrap them to avoid passing keyword argument.

no-changelog-needed

### CODE REVIEW

**Comment by pllim** [comment]
Hello and thanks! Can you please clarify what use cases that this patch is trying to solve? Also, this is technically an API change, so it needs a change log and perhaps some user documentation update as well.

**Comment by nstarman** [comment]
@pllim, sure thing. The general use case is allowing modifications to the Model class construction process in the metaclass, rather than in __new__, which is not the best place for it. For instance, the Model subclasses have the ``inverse`` property overwritten by the metaclass, changing it to ``_inverse`` and using the ``Model.inverse`` function. While this is fine for Model and most subclasses, there is no good way to control this action. Subclasses of Model which wish to exert such type of control need to be able to accept an argument to do so. This may be implemented as a class attribute, or for a more inviolable option, as an arg/kwarg to the class construction itself.

```python
class ExampleModel(Model, metaclass=_ExtendedModelMeta, some_flag=True):
    pass
```

This PR just makes this possible to implement ``_ExtendedModelMeta `` by users foolhardy enough to wade into the murk of metaclasses.

From my experience, I want to override the signature of the evaluate method for a series of subclasses of Polynomial models. Aspects of this process are best / most efficiently accomplished at class declaration, not instantiation. Having flags allows the classes to be powerfully tailored in a manner which may be changed in the child classes and store the choices with inheritable private attributes.

**Comment by nstarman** [comment]
@pllim squashed with tests, change-log, and docs.

**Comment by pllim** [comment]
Thanks. Now let's wait for review from modeling maintainers. In the meantime, we ask for your continued patience.

**Review by nden — APPROVED** [review]

**Comment by nden** [comment]
Looks good

**Comment by nstarman** [comment]
codecov seems to be getting this wrong, none of the coverage changes are related to this PR.

**Review by nden — COMMENTED** [review]
See [comment](https://github.com/astropy/astropy/pull/11098#discussion_r532220641)

**Comment by nstarman** [comment]
> See [comment](https://github.com/astropy/astropy/pull/11098#discussion_r532220641)

Great. I can rebase as necessary. Or vice versa if this is approved first.

File: astropy/modeling/core.py
Comment (embray) [inline astropy/modeling/core.py:73]: Per https://github.com/astropy/astropy/pull/11098#discussion_r532220641 this will have to be rebased on #11098 and remove `__prepare__`.  Alternatively, if this PR is merged first I can rebase mine.  Either way.

File: docs/modeling/new-model.rst
Comment (embray) [inline docs/modeling/new-model.rst:291]: Please wrap this line.

File: astropy/modeling/core.py
Comment (embray) [inline astropy/modeling/core.py:700]: Just to clarify, if you pass keyword arguments to a class, do they get passed to `__init_subclass__` too?  This is just a use case I' haven't seen before.  I would have assumed that if that's the case then the default `__init_subclass__` also takes kwargs but maybe not (hence the need for this).

**Comment by embray** [comment]
I'm still a little confused about this.  What is the *specific* case motivating this?  I'm not aware of any case where it is necessary to subclass `_ModelMeta`, nor do I think the user documentation should say *anything* about extending `_ModelMeta`.

It does seem like accepting some arguments to Model subclasses could be useful in general though.

You mentioned something about the default `_ModelMeta` wrapping `inverse`.  Is there some reason you need it to *not* do that?  What is that reason?  And if so, maybe there is a more directed solution to that problem that can be resolved internally, rather than exposing such machinery at a user-documented level.

File: docs/modeling/new-model.rst
Comment (embray) [reply docs/modeling/new-model.rst:291]: Actually, per [this comment](https://github.com/astropy/astropy/pull/11017#issuecomment-736006874) I don't think this should be documented at all, nor should there be a changelog entry.

If there's a case where you want to be able to prevent wrapping `Model.inverse` maybe that can be an option directly to `_ModelMeta`.  I don't think we should be making anything about `_ModelMeta` publicly documented unless we were to make it a public method in the first place and document explicit use cases for subclassing it.

File: astropy/modeling/core.py
Comment (nstarman) [reply astropy/modeling/core.py]: Thanks! I've rebased. `__prepare__` is gone.

File: docs/modeling/new-model.rst
Comment (nstarman) [reply docs/modeling/new-model.rst:291]: > Actually, per this comment I don't think this should be documented at all, nor should there be a changelog entry.

I agree. I've removed the documentation.

> f there's a case where you want to be able to prevent wrapping Model.inverse maybe that can be an option directly to _ModelMeta.

Maybe for a future PR. 🤞

File: astropy/modeling/core.py
Comment (nstarman) [reply astropy/modeling/core.py:697]: yeah, fortunately or unfortunately the default `__init_subclass__` doesn't take a kwarg. This isn't a problem unless the class has a custom meta.

**Comment by nstarman** [comment]
@embray The *specific* case is something in my personal research.

I want to override the signature of the ``evaluate`` method for a series of subclasses of Polynomial models. Aspects of this process are best / most efficiently accomplished at class declaration, not instantiation. However, when there is a custom metaclass, this type of manipulation is only possible if the metaclass allows for kwargs, which they often do because people (myself including) lazily define the `__new__` and `__init__` methods with `(*args, **kwargs)`. `_ModelMeta` was actually well programmed, which is a problem in this case. I just want to add `**kwargs` to the metaclass. These are non-functional on `_ModelMeta`,  but allow custom subclasses a greater degree of flexibility.

> It does seem like accepting some arguments to Model subclasses could be useful in general though

agreed!

**Comment by embray** [comment]
@nstarman What do you mean by "override the signature of the evaluate method for a series of subclasses of Polynomial models" and is there a reason this couldn't be done with a simple class attribute?

**Comment by nstarman** [comment]
> What do you mean by "override the signature of the evaluate method for a series of subclasses of Polynomial models" and is there a reason this couldn't be done with a simple class attribute?

@embray Sorry for the delay! I kind of lost track of this PR.

I've found a way around this in ``__new__``. What I wanted was to set a flag on the class that it should change the signature off ``evaluate`` when instantiated, while for other subclasses, I did not want that flag set to True. An example of this type of use is https://github.com/adrn/gala/blob/f7184e6b09fbc42a349f6b5a2bca6242f1e9936e/gala/potential/scf/bfe_class.pyx#L53, where the flag ``GSL_only`` is used.
This PR should not change any current functionality or create any new functionality, just enable others to add functionality on top of Model metaclasses.

**Comment by nstarman** [comment]
@nden @embray @perrygreenfield, rebased to fix merge conflict.

**Comment by nstarman** [comment]
@nden @embray @perrygreenfield, the above discussion, while relevant to the reason I pushed this PR, I feel is actually unrelated to the content of this PR, which only enables extending the Model metaclass, and doesn't actually do so here. This PR is certainly not critical for v4.3 and could be pushed to v5.0, but...  #ready-for-final-review ?

**Comment by embray** [comment]
I had a look at that `GSL_only` example, and while I see what that's doing what confuses me is still why this syntax of passing arguments in the `class` statement needs to be used at all in this case, as opposed to the more common syntax of just declaring this in a class attribute like:

```
class SCFPotential(CPotentialBase):
    GSL_only = True
    ...
```

There's certainly nothing *wrong* with the other approach.  It's just less common and seems like overkill to me.  But maybe this is purely an aesthetic/subjective question.  It got me thinking, and I really don't know what a good use case is for custom arguments to `class` so I asked about this here: https://discuss.python.org/t/motivation-for-custom-keyword-arguments-to-class-statement/8402

But maybe there's something I'm completely missing as to why this is necessary.  So I'm definitely not opposed to this PR, and and mostly just struggling to understand the use case.

**Comment by nstarman** [comment]
> But maybe there's something I'm completely missing as to why this is necessary. So I'm definitely not opposed to this PR, and and mostly just struggling to understand the use case.

It's funny because I worked around my use case as well :p ... As I understand it, the limitations of class arguments are twofold:
1. They actually need to be defined *on* the class and can be mutated like any other class method. An argument *to* the class itself need never be stored. It can set a behavior and then be done.
2. clarity: arguments *on* the class can be anything, e.g. ``n_inputs``, while arguments *to* the class are very explicitly arguments to the metaclass constructor.

In the context of  https://discuss.python.org/t/motivation-for-custom-keyword-arguments-to-class-statement/8402, that example is just a flag. But one idea for something that could be done that is more complicated is: wrapping methods for numpy ufuncs versus numba.


```
class ScalarClassWithMethods:
    def __init_subclass__(cls, wrap_method):
        wrapper = numba_wrapper if wrap_method == "numba" else numpy_wrapper
        for name, method in _get_my_methods_:
            setattr(cls, name, wrapper(method))

class NumpyWrapped(ScalarClassWithMethods, wrap_method="numpy"):
    pass

class NumbaWrapped(ScalarClassWithMethods, wrap_method="numba"):
    pass
```

**Comment by pllim** [comment]
Seems like discussions are continuing... I am going to re-milestone to 5.0 so we don't make rash decisions due to lack of time.

**Comment by astrofrog** [comment]
I am re-milestoning to v5.1 as we've now branched v5.0.x

**Comment by pllim** [comment]
Hmm, this was approved by @nden back in Dec 2020 but has been languishing since and yet has a "ready for final review" label. Should the label be removed? What is the current status? Thanks!

**Comment by nstarman** [comment]
Personally, I think it's ready to merge. I might actually remove some of the tests. I'm not sure if the change is large enough to merit testing.

**Review by WilliamJamieson — APPROVED** [review]
These changes seem reasonable to me.

**Comment by pllim** [comment]
Does @nden want to re-review or is an approval from @WilliamJamieson enough?

**Comment by nden** [comment]
William is a maintainer.
Time to merge this.

**Pull request merged**
